discovery/openstack/openstack.go

```diff
@@ -0,0 +1,202 @@
+	openstackLabelInstanceFlavor = openstackLabel + "instance_flavor"
+	openstackLabelTag            = openstackLabel + "tag_"
+)
+
+var (
+	openstackSDRefreshFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_sd_openstack_refresh_failures_total",
+			Help: "The number of OpenStack-SD scrape failures.",
+		})
+	openstackSDRefreshDuration = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Name: "prometheus_sd_openstack_refresh_duration_seconds",
+			Help: "The duration of a OpenStack-SD refresh in seconds.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(openstackSDRefreshFailuresCount)
+	prometheus.MustRegister(openstackSDRefreshDuration)
+}
+
+// Discovery periodically performs OpenStack-SD requests. It implements
+// the TargetProvider interface.
+type Discovery struct {
+	os       *gophercloud.AuthOptions
+	region   string
+	interval time.Duration
+	port     int
+}
+
+// NewDiscovery returns a new OpenStackDiscovery which periodically refreshes its targets.
+func NewDiscovery(conf *config.OpenstackSDConfig) (*Discovery, error) {
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: conf.IdentityEndpoint,
+		Username:         conf.Username,
+		UserID:           conf.UserID,
+		Password:         conf.Password,
+		TenantName:       conf.TenantName,
+		TenantID:         conf.TenantID,
+		DomainName:       conf.DomainName,
+		DomainID:         conf.DomainID,
+	}
+
+	return &Discovery{
+		os:       &opts,
+		region:   conf.Region,
+		interval: time.Duration(conf.RefreshInterval),
+		port:     conf.Port,
+	}, nil
+}
+
+// Run implements the TargetProvider interface.
+func (d *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	ticker := time.NewTicker(d.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := d.refresh()
+	if err != nil {
+		log.Error(err)
+	} else {
+		select {
+		case ch <- []*config.TargetGroup{tg}:
+		case <-ctx.Done():
+			return
+		}
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			tg, err := d.refresh()
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+
+			select {
+			case ch <- []*config.TargetGroup{tg}:
+			case <-ctx.Done():
+				return
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (d *Discovery) refresh() (tg *config.TargetGroup, err error) {
+	t0 := time.Now()
+	defer func() {
+		openstackSDRefreshDuration.Observe(time.Since(t0).Seconds())
+		if err != nil {
+			openstackSDRefreshFailuresCount.Inc()
+		}
+	}()
+	provider, err := openstack.AuthenticatedClient(*d.os)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create openstack session: %s", err)
+	}
+	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Region: d.region,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create openstack compute session: %s", err)
+	}
+
+	opts := servers.ListOpts{}
+	pager := servers.List(client, opts)
+
+	tg = &config.TargetGroup{}
```

Comment:
I think you have to set the `Source` field here, as e.g. in https://github.com/prometheus/prometheus/blob/master/discovery/gce/gce.go#L147-L149

I just noticed that Azure SD does not do that, and I think that is a bug.